src/sinks/nats.rs

```diff
@@ -186,32 +194,34 @@ impl StreamSink<Event> for NatsSink {
 
             let event_byte_size = event.size_of();
 
+            let finalizers = event.take_finalizers();
             let mut bytes = BytesMut::new();
-            if self.encoder.encode(event, &mut bytes).is_err() {
-                // Error is logged by `Encoder`.
-                continue;
-            }
-
-            let byte_size = bytes.len();
+            self.encoder.encode(event, &mut bytes).map_err(|_| {
+                // Error is handled by `Encoder`.
+                finalizers.update_status(EventStatus::Errored);
+            })?;
 
-            match self.connection.publish(&subject, bytes).await {
+            match self.connection.publish(&subject, &bytes).await {
+                Err(error) => {
+                    emit!(NatsEventSendError { error });
+                    finalizers.update_status(EventStatus::Errored);
+                    continue;
+                }
                 Ok(_) => {
+                    finalizers.update_status(EventStatus::Delivered);
+                    self.acker.ack(1);
```

Comment:
I think this move is wrong, or at least needs to go into more places. The `acker` must be updated even if an error happened, otherwise the buffer reader will never advance.

Reply:
Moved the `acker` back out of the `match` so it'll run regardless of success or error 902d5ee7c284281bc01b8c747bf4148f951c4338